Modules/_abc.c

```diff
@@ -0,0 +1,617 @@
+abcmeta_dealloc(abc *tp)
+{
+    Py_CLEAR(tp->abc_registry);
+    Py_CLEAR(tp->abc_cache);
+    Py_CLEAR(tp->abc_negative_cache);
+    Py_TYPE(tp)->tp_free((PyObject *)tp);
+}
+
+static int
+abcmeta_traverse(PyObject *self, visitproc visit, void *arg)
+{
+    Py_VISIT(((abc *)self)->abc_registry);
+    Py_VISIT(((abc *)self)->abc_cache);
+    Py_VISIT(((abc *)self)->abc_negative_cache);
+    return PyType_Type.tp_traverse(self, visit, arg);
+}
+
+static int
+abcmeta_clear(abc *tp)
+{
+    if (tp->abc_registry) {
+        PySet_Clear(tp->abc_registry);
+    }
+    if (tp->abc_cache) {
+        PySet_Clear(tp->abc_cache);
+    }
+    if (tp->abc_negative_cache) {
+        PySet_Clear(tp->abc_negative_cache);
+    }
+    return PyType_Type.tp_clear((PyObject *)tp);
+}
+
+static PyObject *
+abcmeta_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    abc *result = NULL;
+    PyObject *ns, *bases, *items, *abstracts, *base_abstracts;
+    PyObject *key, *value, *item, *iter;
+    Py_ssize_t pos = 0;
+    result = (abc *)PyType_Type.tp_new(type, args, kwds);
+    if (!result) {
+        return NULL;
+    }
+    /* Set up inheritance registry. */
+    result->abc_registry = PySet_New(NULL);
+    result->abc_cache = PySet_New(NULL);
+    result->abc_negative_cache = PySet_New(NULL);
+    if (!result->abc_registry || !result->abc_cache || !result->abc_negative_cache) {
+        Py_DECREF(result);
+        return NULL;
+    }
+    result->abc_negative_cache_version = abc_invalidation_counter;
+    if (!(abstracts = PyFrozenSet_New(NULL))) {
+        Py_DECREF(result);
+        return NULL;
+    }
+
+    /* Compute set of abstract method names in two stages:
+       Stage 1: direct abstract methods.
+       (It is safe to assume everything is fine since type.__new__ succeeded.) */
+    ns = PyTuple_GET_ITEM(args, 2);
+    items = PyMapping_Items(ns); /* TODO: Fast path for exact dicts with PyDict_Next */
+    for (pos = 0; pos < PySequence_Size(items); pos++) {
+        item = PySequence_GetItem(items, pos);
+        key = PyTuple_GetItem(item, 0);
+        value = PyTuple_GetItem(item, 1);
+        int is_abstract = _PyObject_IsAbstract(value);
+        if (is_abstract < 0) {
+            Py_DECREF(item);
+            Py_DECREF(items);
+            goto error;
+        }
+        if (is_abstract && PySet_Add(abstracts, key) < 0) {
+            Py_DECREF(item);
+            Py_DECREF(items);
+            goto error;
+        }
+        Py_DECREF(item);
+    }
+    Py_DECREF(items);
+
+    /* Stage 2: inherited abstract methods. */
+    bases = PyTuple_GET_ITEM(args, 1);
+    for (pos = 0; pos < PyTuple_Size(bases); pos++) {
+        item = PyTuple_GET_ITEM(bases, pos);
+        base_abstracts = _PyObject_GetAttrId(item, &PyId___abstractmethods__);
+        if (!base_abstracts) {
+            if (!PyErr_ExceptionMatches(PyExc_AttributeError)) {
+                goto error;
+            }
+            PyErr_Clear();
+            continue;
```

Comment:
I merged GH-5222.
You can use `_PyObject_LookupAttrId()` after merging master.

Reply:
Great, thanks!